[Security Solution] Allow users to sort Endpoint list by every column

gergoabraham opened the pull request:
## Summary

This PR adds sorting to Endpoint list, it works with all columns.

Details:
- by default it sorts as before: by enrollment date descending
- the sorting is persisted in the URL
- the `GET api/endpoint/metadata` route now accepts `sortField` and `sortDirection` query params, but is quite strict: it only accepts the fields shown in the table for sorting, otherwise returns `400`. reason: feels better to fail on schema validation than allowing any string to go through and returning an internal error
- the API expects the `sortField` to harmonise with the returned `HostInfoInterface` type, so looking from the outside it is consistent. internally, it maps the incoming `sortField` to the internal structure (e.g. `metadata.host.ip` -> `united.endpoint.host.ip`)
- ~**update:** moves `last_checkin` to be a runtime field, so it can be used for sorting~
- ~**update:** adds `Enrolled at` as a column to the table, and `enrolled_at` as a new field in the response~

Todo:
- [x] fix typescript type errors
- [x] add tests

![image](https://github-production-user-asset-6210df.s3.amazonaws.com/39014407/254272727-fe1d69d6-b006-4ad2-9649-966d4131ca9e.png)



### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

paul-tavares commented on x-pack/plugins/security_solution/common/endpoint/types/index.ts:
```diff
@@ -1333,6 +1333,8 @@ interface BaseListResponse<D = unknown> {
   page: number;
   pageSize: number;
   total: number;
+  sortField: string;
```
I don't remember how many of our types uses this base interface, but if its more than just the Endpoint list, then these two additional properties are now showing up on other types, which I don't think is what we want (for now).  you might need to add them explicitly to the metadata list return type.

gergoabraham replied:
it's not used elsewhere. so... I guess it's not even a base interface

just removed one level of the unnecessary abstraction:
```ts
export interface MetadataListResponse {
  data: HostInfo[];
  page: number;
  pageSize: number;
  total: number;
  sortField: string;
  sortDirection: 'asc' | 'desc';
}
```

gergoabraham replied:
ba235c6b59dedeeb7c09225cb257797780563333

paul-tavares replied:
Yeah, the goal with that "base" type was to ensure that going forward our API responses (new APIs) all get a "standard" interface. This is something I remember discussing during a Eng. Hour meeting a while back. 

I'm ok with the change you did (removed it)

paul-tavares commented on x-pack/plugins/security_solution/public/management/pages/endpoint_hosts/view/index.tsx:
```diff
@@ -512,6 +531,12 @@ export const EndpointList = () => {
           onChange={onTableChange}
           loading={loading}
           rowProps={setTableRowProps}
+          sorting={{
```
in order to avoid needless re-renders, I would suggest memoizing the object provided to this `sorting` prop.

gergoabraham replied:
done. also memoized some other variables passed as props, because there were too much re-renders on e.g. opening the details flyout.

there are still a lot of re-renders, but now it doesn't go down to the table that much

eb3413f2e3effd1cd8220624cc43fb0f33b155f4

paul-tavares reviewed (approved):
Did not run it locally, but code looks good. Left two comments for your eval/consideration.

gergoabraham commented:
@elasticmachine merge upstream

ashokaditya commented on x-pack/plugins/security_solution/common/api/endpoint/metadata/list_metadata_route.ts:
```diff
@@ -16,6 +16,21 @@ export const GetMetadataListRequestSchema = {
+        schema.oneOf([
+          // allowed fields for sorting - these are the column fields in the EndpointList table, based on the
+          // returned `HostInfoInterface` data type, and not on the internal data structure
+          schema.literal('metadata.host.hostname'),
+          schema.literal('host_status'),
+          schema.literal('metadata.Endpoint.policy.applied.name'),
+          schema.literal('metadata.Endpoint.policy.applied.status'),
+          schema.literal('metadata.host.os.name'),
+          schema.literal('metadata.host.ip'),
+          schema.literal('metadata.agent.version'),
```
nit: maybe save these literals into an enum and use it here. Would also make it easy to use that in the tests that verify the sort fields.

gergoabraham replied:
done! thanks for the idea 🙌 
70bf06b1bfceece78c9db3de57f405bc05700705

ashokaditya reviewed (approved):
This is awesome. 🔥 Tested it out and works nicely. I have some questions and suggestions. Let me know what you think.

gergoabraham commented on x-pack/plugins/security_solution/server/endpoint/routes/metadata/handlers.ts:
```diff
@@ -54,6 +56,8 @@ export function getMetadataListRequestHandler(
         total,
         page: request.query.page || ENDPOINT_DEFAULT_PAGE,
         pageSize: request.query.pageSize || ENDPOINT_DEFAULT_PAGE_SIZE,
+        sortField: request.query.sortField || ENDPOINT_DEFAULT_SORT_FIELD,
```
that's a good question! I followed the AC in the original issue to use the existing sorting by default, because I found it reasonable to not change the existing behaviour, but extend it with the possibility to sort by other fields. but I agree, it's a bit cumbersome to remove the sorting.

could you help, @caitlinbetz? the problem is that the enrollment date used for default sorting is not displayed in the table, so the user cannot go back to the default sorting easily. what do you think is better for users? keeping the original sorting they are used to (but making it hard to return to), or changing the default sorting, so it's made based on a displayed field?

with a bit more effort, we also can solve this by adding a new column for the 'enrolled' field, but this makes the already packed table even denser, or adding a 'clear sorting' button but that's a bit weird

ashokaditya replied:
I think UX wise it is best that if we're introducing sorting on the table then we start with a sorted column on the table that clearly informs the user what is sorted, and not sort with an invisible UX value. Using `enrolled at` to sort made sense earlier as we were not able to sort any column; with this change I think it is obsolete.

gergoabraham commented on x-pack/plugins/security_solution/public/management/cypress/e2e/mocked_data/endpoints.cy.ts:
```diff
@@ -36,4 +37,75 @@ describe('Endpoints page', () => {
+        expect(body.sortDirection).to.equal('desc');
+      });
+
+      // no sorting indicator is present on screen
```
e821a62f6ca37f53bf2fcff297594c70dc65fda5

gergoabraham commented on x-pack/plugins/security_solution/public/management/cypress/e2e/mocked_data/endpoints.cy.ts:
```diff
@@ -36,4 +37,75 @@ describe('Endpoints page', () => {
   });
+
+  describe('Sorting', () => {
+    it('Sorts by enrollment date descreasing by default', () => {
```
e821a62f6ca37f53bf2fcff297594c70dc65fda5

gergoabraham commented on x-pack/plugins/security_solution/public/management/pages/endpoint_hosts/store/middleware.ts:
```diff
@@ -353,18 +353,25 @@ async function endpointListMiddleware({
 
   try {
     const decodedQuery: Query = searchBarQuery(getState());
+    const sorting = sortField ? { sortField, sortDirection } : {};
```
it turned out that it's perfectly fine, because `coreStart.http.get` filters undefined values from the request. so I simply removed this whole check
3bb8616ae3226cfc34b0a2e9313fa6b7d1f40ce3

gergoabraham commented on x-pack/plugins/security_solution/common/endpoint/types/index.ts:
```diff
@@ -1344,7 +1334,14 @@ export interface AdditionalOnSwitchChangeParams {
+  page: number;
+  pageSize: number;
+  total: number;
+  sortField: string;
```
now it uses the enum

gergoabraham commented on x-pack/plugins/security_solution/public/management/pages/endpoint_hosts/types.ts:
```diff
@@ -97,6 +101,10 @@ export interface EndpointIndexUIQueryParams {
   /** Which page to show */
   page_index?: string;
+  /** Field used for sorting */
+  sort_field?: string;
```
this also uses the new enum

ashokaditya replied:
I don't see the change here on EndpointState. Looks like you made the change to EndpoitIndexUIQueryParams

gergoabraham replied:
🦅 👁️ ! 🙌 
6b5b21985c8760a16ac6f3444536532c41c3780c

gergoabraham commented on x-pack/plugins/security_solution/common/api/endpoint/metadata/list_metadata_route.ts:
```diff
@@ -16,6 +16,21 @@ export const GetMetadataListRequestSchema = {
+          schema.literal('metadata.host.os.name'),
+          schema.literal('metadata.host.ip'),
+          schema.literal('metadata.agent.version'),
+          schema.literal('metadata.@timestamp'),
```
what do you mean by '`last_checkin` or `metadata.@timestamp`'? cannot it be simply `last_checkin`?

gergoabraham replied:
I'm reading the comments in your PR. based on this (https://github.com/elastic/kibana/pull/160506#discussion_r1247892656), would that mean that `last_checkin` is always available and there is no need to fallback to `metadata.@timestamp`?

ashokaditya replied:
Yeah! You can just use the `last_checkin` from host info.

gergoabraham commented:
@elasticmachine merge upstream

ashokaditya commented on x-pack/plugins/security_solution/common/endpoint/types/index.ts:
```diff
@@ -1331,6 +1331,22 @@ export interface AdditionalOnSwitchChangeParams {
+  HOST_OS_NAME = 'metadata.host.os.name',
+  HOST_IP = 'metadata.host.ip',
+  AGENT_VERSION = 'metadata.agent.version',
+  LAST_SEEN = 'metadata.@timestamp',
```
this would now change to `last_checkin`

ashokaditya reviewed (approved):
Thanks for all the changes 🥇 Once we update this to use `last_checkin` this is good to 🚢 . I would also suggest that we introduce `last seen` as the default to be sorted instead of `enrolled at` time. @caitlinbetz @kevinlog @paul-tavares

paul-tavares commented:
@ashokaditya ,

Keep in mind that if we change the default sort order to `last seen` the items on the list will start to jump around (we have auto-refresh on the list) as agents check in, which makes the list a bit erratic. I think we might have had it that way before and then changed it to enroll.

ashokaditya commented:
> @ashokaditya ,
> 
> Keep in mind that if we change the default sort order to `last seen` the items on the list will start to jump around (we have auto-refresh on the list) as agents check in, which makes the list a bit erratic. I think we might have had it that way before and then changed it to enroll.

Ah yes that makes sense. Let's not sort with last checkin time to start with. I was chatting with @kevinlog earlier and he suggests that we should perhaps add the enrolled at column instead.

gergoabraham commented:
@ashokaditya @paul-tavares @kevinlog @caitlinbetz 
I've added the 'Enrolled at' column to the table, so this way:
- we keep the original sort (`enrolled_at: desc`) by default,
- but the user still can change it and return to it in a straightforward way.

The table is now a bit dense, so I tried to improve the sizing of the columns. The width of the columns 'Policy status', 'OS', and 'Actions' is now fixed, as they always have the same content size, while other columns can grow with the size of the window.

Here are 2 screenshots, with a smaller and with a larger resolution:

![image](https://github.com/elastic/kibana/assets/39014407/3ef2f6fe-e526-47f5-b6f9-7553cadd3056)

![image](https://github.com/elastic/kibana/assets/39014407/a17613da-57e3-46ad-a8ac-98938e287211)

Let me know what you think.
@ashokaditya thanks again for your feedback, it improved both the code and the UX a lot! 💯

gergoabraham commented:
@elasticmachine merge upstream

gergoabraham commented:
@elasticmachine merge upstream

paul-tavares commented:
Thanks @gergoabraham for the changes (and @ashokaditya for the detailed review). 

I like the addition of the column to the table - looks good. and yes, I also agree that it is a bit crowded now with the additional column, but am OK with it as is for now. The endpoint list is long overdue for a makeover, we can always try to address the issue then (ex. maybe we move to using a type of card grid, like the artifact list pages. Just a thought for the future.

Thanks again for the work on this.

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/metadata/endpoint_metadata_service.ts:
```diff
@@ -295,7 +297,10 @@ export class EndpointMetadataService {
       },
       last_checkin:
-        _fleetAgent?.last_checkin || new Date(endpointMetadata['@timestamp']).toISOString(),
+        _last_checkin ||
```
So here `_last_checkin` param is redundant.

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/metadata/endpoint_metadata_service.ts:
```diff
@@ -444,8 +449,17 @@ export class EndpointMetadataService {
           ..._agent,
           status: doc?.fields?.status?.[0] as AgentStatus,
         };
+        const lastCheckin = doc?.fields?.last_checkin?.[0] as string | undefined;
```
You can update the related comment above line 445 if you make the change 😅

gergoabraham replied:
done! thanks for the suggestion, Ash!
3bdcf5abe53f23b1d494f1b47107343c53751c9a

gergoabraham replied:
with some additional type safety: bc51a97b4d4c9e3d840bf1c96e2bbb88a882b230

kevinlog commented:
@gergoabraham 

I talked with @caitlinbetz , we think it may be best to not add the Enrolled at column yet.  We understand that it will be strange for the user to have to navigate back to this page to get the original sort, but we think the Enrolled column would be too crowded.  Also, this functionality is very useful and we should merge it.  In the next iteration, we can figure out how we want to introduce the column.

ashokaditya commented on x-pack/plugins/security_solution/server/endpoint/services/metadata/endpoint_metadata_service.ts:
```diff
@@ -442,23 +438,18 @@ export class EndpointMetadataService {
+
+        const runtimeFields = {
+          status: doc?.fields?.status?.[0] as AgentStatus,
+          last_checkin: doc?.fields?.last_checkin?.[0] as string | undefined,
```
I believe the typecasting for `last_checkin` is actually redundant. The field values are already `string | undefined`. We do need the type cast for `status` as they are a union of literals and not `string`

gergoabraham replied:
you're right, it's redundant now, but not because `last_checkin` is string. 😬 

this is the type of `fields`: `fields?: Record<string, any>;`, so the value is `any`, that's why I needed to cast `last_checkin`

but now that `runtimeFields` is `Partial<typeof _agent>`, both type assertions are unnecessary, indeed. so, removed: 9f7e99a5e418358ea10804f87b41a824944e1768

gergoabraham commented:
@elasticmachine merge upstream
